### ISSUE TITLE
How to use Upsert in async environment? (It blocks and causes spin, 100% CPU after log memory is full, or after checkpoint)

### ISSUE THREAD

**Issue by resetcoder** [microsoft/FASTER#355]
I'm using `FasterKV<SpanByte, SpanByte>` (1.7.1) with relatively small log memory (_SegmentSizeBits = 1GB, PageSizeBits = 128kB, MemorySizeBits = 4MB, CheckpointType.Snapshot_), only for test purposes, otherwise I use larger memory settings.

`ReadAsync()`+`Complete()`+`WaitForCommitAsync()` works fine, but since `Upsert()` doesn't have async methods I'm using is synchronously in async function. (I'm aware that there was an `UpsertAsync()` implementation that has been removed, not sure why.)

I have a timer that calls the checkpoint function periodically and creates full snapshot like

```
await kv.TakeFullCheckpointAsync(checkpointSettings.CheckPointType);
await kv.CompleteCheckpointAsync();
```
Checkpointing also works fine, but with the above settings I can insert ~30k entries _(~35 byte key and ~1kB value)_, and then `Upsert()` starts blocking, and [this while loop in AllocatorBase.cs](https://github.com/microsoft/FASTER/blob/master/cs/src/core/Allocator/AllocatorBase.cs#L1460) maxes out the load on all of my CPU cores, and even the next full checkpoint will not make it continue Upserting.

The program just enters the Upsert method and never leaves, all my async tasks get blocked (many concurrent tasks are calling the `ReadAsync()` and `Upsert()` methods.

I tried everything I could find, tried calling `CompletePending()` and `CompletePendingAsync()` before and after in every way I could come up with, I tried calling it while taking snapshot (sometimes I get greater than 0 numbers from `session.GetPendingRequests().Count()` at taking checkpoint, calling any completependings didn't help then either).

How do I use `Upsert()` properly in this environment, or how do I create checkpoint/commit properly, to avoid blocking on `Upsert()`? I see there is `RMWAsync()`, is it possible to use it for inserting and updating? _(I tried it and it created an entry but the value was missing, I may check it out again.)_

Any help will be appreciated!

**Comment by resetcoder** [comment]
Also getting this exception quite often:

```
Exception: System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at FASTER.core.FasterKV`2.HandleOperationStatus[Input,Output,Context,FasterSession](FasterExecutionContext`3 opCtx, FasterExecutionContext`3 currentCtx, PendingContext`3& pendingContext, FasterSession fasterSession, OperationStatus status, Boolean asyncOp, AsyncIOContext`2& request) in ...FASTER\core\Index\FASTER\FASTERImpl.cs:line 1561
```

And this:

```
Exception: System.Exception: CompletePendingAsync did not complete
   at FASTER.core.FasterKV`2.CompletePendingAsync[Input,Output,Context,Functions](ClientSession`6 clientSession, CancellationToken token) in ...FASTER\core\ClientSession\FASTERAsync.cs:line 97
   at FASTER.core.ClientSession`6.CompletePendingAsync(Boolean waitForCommit, CancellationToken token) in ...FASTER\core\ClientSession\ClientSession.cs:line 490
   at FASTER.core.ClientSession`6.WaitForCommitAsync(CancellationToken token) in ...FASTER\core\ClientSession\ClientSession.cs:line 518
```

**Comment by badrishc** [comment]
`Upsert` never returns a status of pending, so UpsertAsync is not necessary. We always insert at the tail of the log and if you have sufficient SSD bandwidth to handle the checkpointing, then it will not spin trying to allocate on the tail of the log. The spin does a Thread.Yield so it will not starve other tasks. Make sure you are calling `TakeHybridLogCheckpointAsync` from a separate thread.

Also, if you want to use `SpanByte`, either use the latest from master or wait for the next release (1.7.2) that fixes some issues with it.

I would recommend, first get your pipeline working properly with `long` keys and values, make sure it is stable and operating well, then switch to `Memory<byte>` or `SpanByte`.

**Comment by badrishc** [comment]
You will need to post a simplified stand-alone repro here of your code that exhibits these exception, then it will be very easy to track and fix them, thanks.

**Comment by badrishc** [comment]
> Operations that change non-concurrent collections must have exclusive access.

This looks like you are trying to use the same session from more than one thread at the same time. That is not allowed. Each session is mono-threaded, i.e., you should not have two tasks or threads operating on the same session at the same instant.

**Comment by badrishc** [comment]
If you are using `ReadAsync`, there is no reason for you to ever call `CompletePending` or `CompletePendingAsync`. These calls are used only when you use synchronous `Read` and it returns a `Status.PENDING` result.

**Comment by resetcoder** [comment]
> if you have sufficient SSD bandwidth to handle the checkpointing
I expected this to be a problem so I moved the log folder to an M.2 SSD, 3.2/2.8GB/s bandwidth, didn't help.

>  Make sure you are calling TakeHybridLogCheckpointAsync from a separate thread.
Checkpointing is called from a System.Timers.Timer async callback function, so it is a task, the thread id probably varies, I will dedicate a thread for this then.

> either use the latest from master
I'm using the latest source.

> working properly with long keys and values
I will try that in order after I made a new thread for checkpointing.

> You will need to post a simplified stand-alone repro here
That will not be easy, this is inside now a webserver, I'm not sure I can extract, using a lot of dependencies, but I will try. I most definitely can not simulate the hundreds of tasks calling the database methods, because they are coming from the async sockets, but I can create some to stress-test, not sure it will have the same effect.

> This looks like you are trying to use the same session from more than one thread at the same time. 
I'm having an ObjectPool object that generates sessions, I rent one for a session-method call, then I return to the pool, never call multiple session-methods in a row. If there is no pending operation on a returned session then I can not rent one again with pending operation, but I will check, if I can use `session.GetPendingRequests()` to check if there is a pending request when I get the session from the pool I can iterate to another with no pending request, I definitely will try this now.

Thank You for the advice, I try these now, and if still get the error, then I try make a simplified repo.

**Comment by badrishc** [comment]
Thank you. Your scenario is quite interesting to us. You are probably hitting some rare corner cases -- with a stress-test repro these should be easy to diagnose. It might also be a good contribution back to our community.

**Comment by badrishc** [comment]
If you are using `ReadAsync` you should make sure you do this sequence:
* Rent a session
* var result = await ReadAsync()
* result.Complete()
* return session to pool --> so that no other task can use this session

For parallelism within a session:

* Rent a session
* task[0] = ReadAsync()
* task[1] = ReadAsync()
...
* foreach task: { (await task).Complete() }
* return session to pool --> so that no other task can use this session until this batch is done

**Comment by badrishc** [comment]
FYI, this PR handles a rare concurrency on ctx.ioPendingReads dictionary: https://github.com/microsoft/FASTER/pull/356

I don't think you were running into this - it happens only in case of disk access error.

**Comment by resetcoder** [comment]
So... Right now I created a dedicated maintenance thread that calls checkpointing, compaction, and searches for expired entries to delete (this is not important, it is my implementation), and cleaned up the ReadAsync/Upsert/RMWAsync/Delete methods as You suggested, and I'm kind of having some success, Thank You!

Now I can insert/update not only ~30k entries as before, now I could insert all the 1 million, that I'm using for this test, which is great, now I'm not running into CPU spinning loops, everything works fine, except when around 600k entries compaction begins, then  the program hangs, CPU load 100% and doesn't get over the `kv.Log.Compact(kv.Log.SafeReadOnlyAddress, true)` call.

Compaction runs in the dedicated thread.

The sessionpool works great, no shared session between tasks/threads. I try dig deeper why compaction causes this hanging.

**Comment by badrishc** [comment]
Log compaction works by creating a in-memory instance of FASTER that will hold quite a few of the live objects, and then they will be written to the tail of the log. So you might probably not want to compact all the way from begin address to `SafeReadOnlyAddress` -- that will incur a full scan. Typically, we suggest compacting around 10-20% of the log at a time.

Also see this info:

You can compact the log from BeginAddress to any address (say NEWBEGIN) you want on disk. FASTER will find the the live records in that region and write them to the tail in memory. To avoid data loss in memory, you can call Compact with parameter shiftBeginAddress = false. This will ensure we do NOT automatically shift the begin address after compaction to tail. Then, you take a checkpoint to ensure recoverability of the newly copied data (foldover is fine). Finally, you call ShiftBeginAddress to address NEWBEGIN.

See comment here for a bit more info: https://github.com/microsoft/FASTER/blob/f35a709f94addfd599d19205f05022c3c73493b3/cs/src/core/Index/FASTER/LogAccessor.cs#L199

**Comment by badrishc** [comment]
Not that log compaction will delete expired entries in the log, i.e., records that are no longer the value for a live key in the database. If you simply want to truncate the log from its begin, you can directly call ShiftBeginAddress. That will not perform actual compaction.

**Comment by resetcoder** [comment]
Thank You for the info, it helps a lot. I've run some tests and I'm still gettings CPU spin, but very rarely, probably still some session/task/thread issue. I need to inspect the all code of this module, but not now, probably in the next 3 days, and I also try to implement new compaction procedure based on Your info. I'll add news in this issue when I'm debugging again. Also, waiting for the Memory<> type. Thanks again for the help!

**Comment by badrishc** [comment]
Latest release 1.7.2 includes memory sample. See here as well: https://github.com/microsoft/FASTER/blob/master/cs/samples/StoreVarLenTypes/Program.cs

SpanByte will be slightly more efficient overall, though.

Closing this issue as resolved, please reopen if you find any issue with compaction.

**Comment by resetcoder** [comment]
@badrishc 
> Note that log compaction will delete expired entries in the log, i.e., records that are no longer the value for a live key in the database.

Is there any way to know while iterating through indices which key is inactive (was overwritten, therefore the active was moved to the tail)?

I mean, if I insert an entry, and then I update it a few times, the live entry will be at the tail, but if I iterate through the log how I know which entry is not live anymore? **RecordInfo.Tombstone is false** and **RecordInfo.Final is true** and **RecordInfo.Invalid is false** on all entries, so it is quite hard to determine what is the difference between them. RecordInfo.Version is incrementing, but it is not really helping, since I don't know which is the last, only if I keep a record somewhere about all keys with versions, but then iterating kind of loses its point in this situation. I guess after compaction the old entries will be removed, but how do I know how many not-live keys do I have?

It would be nice if I could query the log about how many active, inactive and deleted keys I have without iterating every time I need this information. Log after initiation (or recovery) should keep a counter about the situation, and should update realtime, this would help me decide when to start a compaction (like when I have inactive/deleted entries that is over a specific limit).

Knowing which entry is invalid would be also nice, when I'm iterating through the entries (just because I'm searching, or just want to have all my active entries), so that I can ignore the inactive values.

Now I'm trying a different implementation about dispatching the actions to dedicated threads and do a stress test, if I'm running into troubles again I try make a repo, but I can not promise that, since as I said this is working inside a project, relying on custom dependencies that I should not share unfortunately. I do my best to separate it from the project, if needed.

**Comment by resetcoder** [comment]
Just a little follow up: Using the 1.7.2 version and creating a strict work dispatcher (having configurable count of dedicated threads, sessions assigned to each thread, and running only inside the loop of the dedicated thread) fixed all issues.

At the end, when I was calling the dispatcher (which actually is blocking and synchronous) from asynchronous tasks, after a certain amount of concurrent calls the program stalled in deadlock on the call of `kv.CompleteCheckpointAsync().GetAwaiter().GetResult()` method, because even though it is called synchronously, the taskscheduler tried to schedule it, and couldn't because all other tasks in the stress-test (more than 10000) was blocked by the dispatcher.

The solution was, I moved away from `BlockingCollection<T>`  and ported the dispatcher to `BufferBlock<T>`, because it can wait for consumables asynchronously, this made me able to create a completly awaitable async API for the dispatcher without any blocking calls, since that I'm not having any deadlocks, and all functions are working without any issue, stress-tested with thousands of concurrent tasks.

Bottom line: A session should be called only from one thread, maintenance methods (recover, checkpointing, compaction) should have an own dedicated thread, and as always, never block in async calls, because checkpointing will result in a deadlock.

**Comment by badrishc** [comment]
To iterate all distinct keys in the database, you can use this: https://github.com/microsoft/FASTER/pull/287

var iterator = store.Iterate();
